### PR TITLE
add Apa102Pixel struct and FastLED's pseudo-13-bit gamma correction algorithm

### PR DESCRIPTION
**Pull request by Be-ing**
`impl From<RGB8>` is implemented for this new struct and the trait bound for `SmartLedsWrite::write` was already specified as `Into<Self::Color>`, so this is not a breaking change for downstreams passing iterators of RGB8 into `SmartLedsWrite::write`.

Implemented on top of #14 so this change applies to both sync and async implementations.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -11,10 +11,43 @@
...
+    pub r: u8,
+    pub g: u8,
+    pub b: u8,
+    pub brightness: u8,
+}
+
+impl From<RGB8> for RGB8Brightness5 {
+    fn from(old: RGB8) -> Self {
+        // TODO: implement pseudo 13-bit color mixing & gamma correction algorithm from FastLED
+        // https://github.com/FastLED/FastLED/blob/fd00e9e856b758d561b25ddb1ca67c2ee0dbbbc7/src/five_bit_hd_gamma.cpp
```
Comment (Be-ing) [inline src/lib.rs]: FWIW this is why I decided to use APA102 for my project https://old.reddit.com/r/FastLED/comments/1d1wyzp/fastled_370_release/l67z293/

But I don't want to write C++, so I'll rewrite it in Rust :grin:

File: src/pixel.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+/// Brightness is represented by a `u8` without any checks for valid values to make this struct a
+/// zero-cost abstraction. Any `u8` values above the maximum for 5 bits (0b00011111 in binary or 31 in decimal)
+/// will be truncated to the maximum when writing data to the APA102 LEDs.
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub struct Apa102Pixel {
+    pub r: u8,
+    pub g: u8,
+    pub b: u8,
+    pub brightness: u8,
```
Comment (Be-ing) [inline src/pixel.rs]: Using simply `b` to abbreviate `blue` makes it somewhat easy to confuse with `brightness`. We could instead explicitly use `red`, `green`, and `blue`, instead of `r`, `g`, and `b`, though that wouldn't be consistent with the structs of the `rgb` crate... thoughts?

Comment (Be-ing) [reply src/pixel.rs]: Changed to use `red`, `green`, and `blue` explicitly to avoid confusion.

File: src/lib.rs (hunk 1 above)
Comment (Be-ing) [reply src/lib.rs]: I [contributed](https://github.com/FastLED/FastLED/pull/1841) [tests](https://github.com/FastLED/FastLED/pull/1842) to the original C++ implementation to verify that this translated Rust code behaves identically.

File: src/bitshift.rs [hunk 3]
```diff
@@ -0,0 +1,265 @@
+// Manually translated to Rust from FastLED's MIT licensed C++ code
+
+/// Steal brightness from brightness_src and give it to brightness_dst.
+/// After this function concludes the multiplication of brightness_dst and brightness_src will remain constant.
+pub(crate) fn brightness_bitshifter8(
```
Comment (sajattack) [inline src/bitshift.rs:5]: This seems like the sort of thing that would be better served as an external effects crate or similar. I don't think this crate is the right place for it. Does it only apply to apa102 over spi?

Comment (Be-ing) [reply src/bitshift.rs:5]: Yes, this is specific to the APA102 protocol, so I think it belongs in this crate.

**Comment by Be-ing** [comment]
Thank you @sajattack for the hint about the uX crate. Its `u5` type makes sense for the Apa102Pixel struct's `brightness` field.

File: src/asynch.rs [hunk 4]
```diff
@@ -63,7 +59,7 @@ impl<SPI> SmartLedsWriteAsync for Apa102Async<SPI>
 where
     SPI: SpiBus,
 {
-    type Color = RGB8;
```
Comment (david-sawatzke) [inline src/asynch.rs:66]: Could you also retain the old implementation so a user can keep on directly using RGB8. Would also make this PR a non-breaking change

File: src/math.rs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+///
+/// For example, a value representing 40% as a `u16` would be `26,214 / 65,535`.
+/// Converting that to a `u8` would return `102 / 255`, exactly 40% through the
+/// range of values representable by a `u8`.
+pub(crate) fn map16_to_8(x: u16) -> u8 {
+    // Tested to be nearly identical to double precision floating point
+    // doing this operation.
+    if x == 0 {
+        return 0;
+    }
```
Comment (david-sawatzke) [inline src/math.rs]: Why is this a special case? Does ((0 + 128) >> 8) not result in the same thing?

File: src/psuedo13.rs [hunk 6]
```diff
@@ -0,0 +1,215 @@
+// Manually translated to Rust from FastLED's MIT licensed C++ code
```
Comment (david-sawatzke) [inline src/psuedo13.rs]: s/psuedo/pseudo (this file and its references)

**Review by david-sawatzke — COMMENTED** [review]
Could you also add a short code example how one would use this feature, i.e. how to get from an existing RGB8 iterator to `write`. And an introduction as to what this feature does. Ideally in the crate comment in lb.rs? (Including the downside of the lower PWM frequency https://cpldcpu.com/2014/08/27/apa102/)

How much of the code is 1-to-1 from FastLED, should we go to only MIT?

This looks good to me otherwise, sorry for the delay in reviewing

File: src/asynch.rs (hunk 4 above)
Comment (Be-ing) [reply src/asynch.rs:66]: `impl From<RGB8> for Apa102Pixel` is implemented and the trait bound for `SmartLedsWrite::write` was already specified as `Into<Self::Color>`, so this is not a breaking change for downstreams passing iterators of RGB8 into `SmartLedsWrite::write`. Passing RGB8's should behave the same as before.

File: src/math.rs (hunk 5 above)
Comment (Be-ing) [reply src/math.rs]: You're right, there's no need for this special case handling 0.

File: src/asynch.rs (hunk 4 above)
Comment (david-sawatzke) [reply src/asynch.rs:66]: This is unfortunately not equivalent, as while `RGB8` continues to work, other types which implement `Into<RGB8> ` will not as `Into` and `From` are not transitive.

**Comment by Be-ing** [comment]
> Including the downside of the lower PWM frequency https://cpldcpu.com/2014/08/27/apa102/

This is the first I've heard about flickering when using low brightness values on APA102. It is surprising to me because I have not noticed this and I mostly use low brightness. I tested at minimum brightness (1) with my LED strips and, following [this comment](https://cpldcpu.com/2014/08/27/apa102/#comment-521), moving my head, moving the LEDs, and putting the LEDs in my peripheral vision. Doing all of those at the same time, I still can't notice even slight flickering. It's possible that the APA102 LEDs I bought in 2024 are not identical to the ones tested for that blog post in 2014. I found a [video](https://community.particle.io/t/smartmatrix-apa102-library-open-hardware-photon-apa102-shield/21562/74) demonstrating the flickering with a slow motion video, and I don't see anything like that.

I'm surprised this library previously hardcoded maximum brightness. At least on the LEDs I have, maximum brightness is *extremely* bright. So bright that it's painful to look at directly and my 1 meter of 144 LEDs illuminates my whole bedroom.

File: src/asynch.rs (hunk 4 above)
Comment (Be-ing) [reply src/asynch.rs:66]: Ah, good point. If users aren't using RGB8 directly, but something that also `impl From<RGB8>`, they'll need to add a `.into()` (or `.map(Into::into)`) on their side before passing to `write`. But I suspect almost all users have been providing data as `[RGB8]`, so it shouldn't require any code changes for most library users.

**Comment by Be-ing** [comment]
> How much of the code is 1-to-1 from FastLED, should we go to only MIT?

All of the `math`, `bitshift`, and `pseudo13` modules are from FastLED. I copied and pasted the C++ code then rewrote it line-by-line to get it to build in Rust.

IANAL, but I'm confident there is no need to change the license of this library. The MIT license places no restrictions on doing anything with the code so long as attribution is retained.

File: src/asynch.rs (hunk 4 above)
Comment (Be-ing) [reply src/asynch.rs:66]: I have added a doctest to lib.rs (run it with `cargo test --doc`) demonstrating that the old way of passing `[RGB8]` continues to work with no change required for the downstream library user.

**Comment by Be-ing** [comment]
ping

**Comment by Be-ing** [comment]
Following up on this, I'd like to get this merged.

**Comment by zackees** [comment]
Algorithm inventor, this author added additional unit tests to cover all corner cases

This algorithm unlocks the actual potential of the APA102 and SK9822 chipsets.

Now it's in rust.

**Comment by sajattack** [comment]
Not sure where David's at. Looks like his last github activity was over a month ago. Nevertheless, it seems he was mostly happy with it, aside from some nits you've addressed, so I think this is good to merge.

**Comment by sajattack** [comment]
@Be-ing do you think this warrants a new release on it's own or should we wait for some of your other draft PRs?

**Comment by Be-ing** [comment]
Thank you for merging @sajattack. This could warrant a release on its own, but the other pending PRs are ready for review now, so might as well release them all together.

**Comment by david-sawatzke** [comment]
Thank you for merging and taking a look at everything @sajattack and sorry for not keeping up with Github. Thank you @Be-ing for all the improvements!
